dbms/src/Storages/StorageDeltaMerge.cpp

```diff
@@ -726,13 +730,104 @@ DM::RSOperatorPtr StorageDeltaMerge::parseRoughSetFilter(const SelectQueryInfo &
             };
             rs_operator = FilterParser::parseDAGQuery(*query_info.dag_query, columns_to_read, std::move(create_attr_by_column_id), log);
         }
-        if (likely(rs_operator != DM::EMPTY_FILTER))
+        if (likely(rs_operator != DM::EMPTY_RS_OPERATOR))
             LOG_DEBUG(tracing_logger, "Rough set filter: {}", rs_operator->toDebugString());
     }
     else
         LOG_DEBUG(tracing_logger, "Rough set filter is disabled.");
 
-    return rs_operator;
+    // build push down filter
+    if (likely(query_info.dag_query) && !query_info.dag_query->pushed_down_filters.empty())
+    {
+        NamesAndTypes columns_to_read_name_and_type;
+        for (const auto & col : columns_to_read)
+        {
+            columns_to_read_name_and_type.emplace_back(col.name, col.type);
+        }
+
+        std::unordered_set<String> filter_column_names;
+        for (const auto & filter : query_info.dag_query->pushed_down_filters)
+        {
+            DB::getColumnNamesFromExpr(filter, columns_to_read_name_and_type, filter_column_names);
+        }
+        ColumnDefines filter_columns;
+        filter_columns.reserve(filter_column_names.size());
+        for (const auto & name : filter_column_names)
+        {
+            auto iter = std::find_if(
+                columns_to_read.begin(),
+                columns_to_read.end(),
+                [&name](const ColumnDefine & d) -> bool { return d.name == name; });
+            RUNTIME_CHECK(iter != columns_to_read.end());
+            filter_columns.push_back(*iter);
+        }
+
+        ColumnInfos table_scan_column_info;
+        table_scan_column_info.reserve(columns_to_read.size());
+        const auto & table_infos = tidb_table_info.columns;
+        for (const auto & col : columns_to_read)
+        {
+            auto iter = std::find_if(
+                table_infos.begin(),
+                table_infos.end(),
+                [col](const ColumnInfo & c) -> bool { return c.id == col.id; });
+            RUNTIME_CHECK(iter != table_infos.end());
+            table_scan_column_info.push_back(*iter);
```

Comment:
Why need to use this way to fill `table_scan_column_info`? I think the column order in `tidb_table_info.columns` and `columns_to_read` should be the same?

Reply:
```cpp
for (const auto & col : columns_to_read)
{
    auto iter = std::find_if(
        table_infos.begin(),
        table_infos.end(),
        [col](const ColumnInfo & c) -> bool { return c.id == col.id; });
    RUNTIME_CHECK(iter != table_infos.end());
    table_scan_column_info.push_back(*iter);
}
```

The loop iterator as the order of `columns_to_read`, so it will be the same.

Reply:
oh, I get, I will optimize this.

Reply:
```cpp
ColumnInfos table_scan_column_info;
table_scan_column_info.reserve(columns_to_read.size());
const auto & table_infos = tidb_table_info.columns;
auto table_info_begin = table_infos.begin();
for (const auto & col : columns_to_read)
{
    auto iter = std::find_if(
        table_info_begin,
        table_infos.end(),
        [col](const ColumnInfo & c) -> bool { return c.id == col.id; });
    RUNTIME_CHECK(iter != table_infos.end());
    table_info_begin = iter + 1;
    table_scan_column_info.push_back(*iter);
}
```